admin/tabs/roles/RolesModel.ts

```diff
@@ -0,0 +1,280 @@
+import {RoleEditorModel} from '@xh/hoist/admin/tabs/roles/editor/RoleEditorModel';
+import {HoistRole} from '@xh/hoist/admin/tabs/roles/HoistRole';
+import {FilterChooserModel} from '@xh/hoist/cmp/filter';
+import {GridModel} from '@xh/hoist/cmp/grid';
+import * as Col from '@xh/hoist/cmp/grid/columns';
+import {div, li, ul} from '@xh/hoist/cmp/layout';
+import {HoistModel, LoadSpec, managed, persist, ReactionSpec, XH} from '@xh/hoist/core';
+import {RecordActionSpec} from '@xh/hoist/data';
+import {fmtDate} from '@xh/hoist/format';
+import {Icon} from '@xh/hoist/icon';
+import {bindable, makeObservable} from '@xh/hoist/mobx';
+import {pluralize} from '@xh/hoist/utils/js';
+import {compact, groupBy, mapValues} from 'lodash';
+import {action, observable} from 'mobx';
+import moment from 'moment/moment';
+
+export class RolesModel extends HoistModel {
+    static PERSIST_WITH = {localStorageKey: 'xhAdminRolesState'};
+
+    override persistWith = RolesModel.PERSIST_WITH;
+
+    readonly ACTIONS: RecordActionSpec[] = this.createRecordActions();
+
+    @managed readonly gridModel: GridModel = this.createGridModel();
+    @managed readonly filterChooserModel: FilterChooserModel = this.createFilterChooserModel();
+    @managed readonly roleEditorModel = new RoleEditorModel(this);
+
+    @observable.ref allRoles: HoistRole[] = [];
+
+    @bindable @persist groupByCategory = true;
+
+    get selectedRole(): HoistRole {
+        return this.gridModel.selectedRecord?.data as HoistRole;
+    }
+
+    constructor() {
+        super();
+        makeObservable(this);
+        this.addReaction(this.groupByCategoryReaction());
+    }
+
+    override async doLoadAsync(loadSpec: LoadSpec) {
+        try {
+            const {data} = await XH.fetchJson({loadSpec, url: 'roleAdmin/list'});
+            if (loadSpec.isStale) return;
+            this.setRoles(this.processRolesFromServer(data));
+        } catch (e) {
+            if (loadSpec.isStale) return;
+            XH.handleException(e);
+            this.clear();
```

Comment:
I have a longstanding item in my mind to try and work through the best patterns for doLoadAsync() w.r.t. exception handling / reporting and autorefresh.  The issue here is that an autorefresh that fails can silently clear the grid and you end up with a "clean" looking UI that reports no roles.

I'd suggest also returning as a no-op if `loadSpec.isAutorefresh` - happy to discuss some more thoughts around use of `lastLoadException`, but that is where it gets a bit bogged down.